iOS/TvShows/TvShows/Common/Network/BaseApiClient.swift

```diff
@@ -0,0 +1,164 @@
+    }
+
+    func get<ResultType: Decodable>(
+        path: String,
+        queryParameters: [String: String]?
+    ) -> Single<ResultType> {
+        executeAndReturn(path: url(with: path), method: .get, parameters: queryParameters)
+    }
+
+    func post<ParamsType: Encodable, ResultType: Decodable>(
+        path: String,
+        body: ParamsType
+    ) -> Single<ResultType> {
+        executeAndReturn(path: url(with: path), method: .post, parameters: body)
+    }
+
+    private func executeAndReturn<ParamsType: Encodable, ResultType: Decodable>(
+        path: String,
+        method: HTTPMethod,
+        parameters: ParamsType? = nil
+    ) -> Single<ResultType> {
+        .create { [weak self] single in
+            guard let self = self else { return Disposables.create() }
+
+            let request = self.buildRequest(path: path, method: method, parameters: parameters)
+            self.urlSession.dataTask(with: request) { [weak self] data, response, error in
+                guard
+                    let httpReponse = response as? HTTPURLResponse,
+                    let self = self,
+                    let statusCode = HttpStatusCode(rawValue: httpReponse.statusCode)
+                else {
+                    single(.failure(ApiError.badRequest))
+                    return
+                }
+                
+                if let error = self.mapToApiError(status: statusCode) {
+                    single(.failure(error))
+                    return
+                }
+                
+                guard let result: ResultType = self.parse(data: data) else {
+                    single(.failure(ApiError.noData))
+                    return
+                }
+
+                single(.success(result))
+            }
+            .resume()
+            
+            return Disposables.create()
+        }
+    }
+
+    private func execute<ParamsType: Encodable>(
+        path: String,
+        method: HTTPMethod,
+        parameters: ParamsType? = nil
+    ) -> Completable {
+        .create { [weak self] observer  in
+            guard let self = self else { return Disposables.create() }
+
+            let request = self.buildRequest(path: path, method: method, parameters: parameters)
+            self.urlSession.dataTask(with: request) { [weak self] _, response, error in
+                guard
+                    let httpReponse = response as? HTTPURLResponse,
+                    let self = self,
+                    let statusCode = HttpStatusCode(rawValue: httpReponse.statusCode)
+                else {
+                    observer(.error(ApiError.badRequest))
+                    return
+                }
+
+                if let error = self.mapToApiError(status: statusCode) {
+                    observer(.error(error))
+                    return
+                }
+
+                observer(.completed)
+
+            }
+            .resume()
+
+            return Disposables.create()
```

Comment:
There is some room for improvement here regarding the cancelation of the url task.